### PR TITLE
Reconcile CAPC Cluster as CloudStack CKS Cluster (Default: Opt Out)

### PR DESCRIPTION
**Pull request by rohityadavcloud**
*Issue #, if available:*

https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack/issues/310

*Description of changes:*

The notion of a CAPC cluster represents a k8s cluster created and managed by CAPC. While a similar construct exists in CloudStack in form of CKS (CloudStack K8S Service) cluster. This PR aims to address #310 by introducing logic that can reconcile a CAPC cluster to a CKS cluster, while having this feature disabled by default (i.e. opt out by default). This will be useful for users (and integrations) who want to be aware about a potentially externally managed K8S cluster in CloudStack via the reconciled CKS cluster.

ACS 4.19 and later support such reconciliation with introduction of feature support PRs in upstream CloudStack releases:
https://github.com/apache/cloudstack-go/pull/59
https://github.com/apache/cloudstack/pull/7515
https://github.com/apache/cloudstack-documentation/pull/315

Previously attempted as https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack/pull/250

*Testing performed:*

To be updated

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by rohityadavcloud** [comment]
Currently - in progress, draft - not fit or ready for review yet.

**Comment by vishesh92** [comment]
@g-gaston Please review this PR. I am targeting this PR for v0.5.0.

**Review by weizhouapache — APPROVED** [review]
overall lgtm

**Comment by rohityadavcloud** [comment]
@g-gaston this looks like is ready for merging, the tests have passed. We'll keep this open for another week if you're able to help review this, thanks.

**Comment by g-gaston** [comment]
> @g-gaston this looks like is ready for merging, the tests have passed. We'll keep this open for another week if you're able to help review this, thanks.

Apologies for the delay! Yes i will get to this next week

**Comment by g-gaston** [comment]
Sorry again for the delay!

The API and implementation logic makes sense, however I would love to talk about the controller architecture. I would like to propose to implement this in a separate controller/s instead of adding extra if/else condition to the existing controllers.

IMO this has several advantages:
* It decouples this new feature from the critical path for the normal cluster reconciliation path. That's the main responsibility of this provider and the one we should prioritize among all others. In practice, it reduces the chances of introducing issues/bugs in that path and it avoids extra latency.
* It keeps a cleaner separation of concerns, so the main Cluster/Machine controllers only care about the CAPI provider responsibilities. IMO this makes the code cleaner and easier to follow.
* It makes this functionality modular: it opens the possibility of running this functionality outside of the CAPC controller manager, as something that builds on top of CAPC and not necessarily part of CAPC. It also allows cluster admins to enable/disable this functionality at the controller manager level, which for a new feature is quite useful.

Opinions?

**Comment by vishesh92** [comment]
> Sorry again for the delay!
> 
> The API and implementation logic makes sense, however I would love to talk about the controller architecture. I would like to propose to implement this in a separate controller/s instead of adding extra if/else condition to the existing controllers.
> 
> IMO this has several advantages:
> 
>     * It decouples this new feature from the critical path for the normal cluster reconciliation path. That's the main responsibility of this provider and the one we should prioritize among all others. In practice, it reduces the chances of introducing issues/bugs in that path and it avoids extra latency.
> 
>     * It keeps a cleaner separation of concerns, so the main Cluster/Machine controllers only care about the CAPI provider responsibilities. IMO this makes the code cleaner and easier to follow.
> 
>     * It makes this functionality modular: it opens the possibility of running this functionality outside of the CAPC controller manager, as something that builds on top of CAPC and not necessarily part of CAPC. It also allows cluster admins to enable/disable this functionality at the controller manager level, which for a new feature is quite useful.
> 
> 
> Opinions?

@g-gaston Makes sense. I have done the changes you requested. Please review again.

File: api/v1beta1/conversion.go
Comment (g-gaston) [inline api/v1beta1/conversion.go:54]: why do we need another sprintf? can't we use Errorf with a %v?

File: api/v1beta3/cloudstackcluster_types.go
Comment (g-gaston) [inline api/v1beta3/cloudstackcluster_types.go:40]: nit: I could be wrong bu I thought the recommended way was to use pointers for optional fields? I think that's what I remember from the k8s api conventions

File: controllers/cks_cluster_controller.go
Comment (g-gaston) [inline controllers/cks_cluster_controller.go:73]: nit: in go is more common to express this as a guard condition
```suggestion
	if !r.CSCluster.Spec.SyncWithACS {
	  	return ctrl.Result{}, nil
	}
```

It has the added benefit of reducing the indentation of the main business logic and imo makes reading it easier since you don't have to hold the if condition in your head

File: controllers/cks_cluster_controller.go
Comment (g-gaston) [inline controllers/cks_cluster_controller.go:81]: can we guarantee all clustes have at least one failure domain? else this could panic

File: controllers/cks_cluster_controller.go
Comment (g-gaston) [inline controllers/cks_cluster_controller.go:91]: should we just return an error? if the plugin is not enabled in CS but the cluster is configured to use CKS cluster, this is a configuration error, right? or at least log it as an error

File: controllers/cks_cluster_controller.go
Comment (g-gaston) [inline controllers/cks_cluster_controller.go:93]: logr favors using key values instead of string interpolation. also wouldn't controller runtime log this error if we are returning an error?
```suggestion
			r.Log.Info("Failed creating ExternalManaged kubernetes cluster on CloudStack", "error", err.Error()))
```
I think Log also has log.Error

File: controllers/cks_cluster_controller_test.go
Comment (g-gaston) [inline controllers/cks_cluster_controller_test.go:50]: test for delete?

File: controllers/cks_machine_controller.go
Comment (g-gaston) [inline controllers/cks_machine_controller.go:69]: same here about the guard

File: controllers/cks_machine_controller.go
Comment (g-gaston) [inline controllers/cks_machine_controller.go:68]: do we not want a finalizer on the CS machines as well?

File: controllers/cloudstackmachine_controller_test.go
Comment (g-gaston) [inline controllers/cloudstackmachine_controller_test.go:58]: do we need this?

File: main.go
Comment (g-gaston) [inline main.go:277]: could we make this optional? either a flag or some other type of config that enables these controllers
And IMO it should be opt-in for now, so users who are not interested on this don't run these controllers by default

**Review by g-gaston — COMMENTED** [review]
Thanks so much for the refactor, looking great!

File: controllers/cks_cluster_controller.go
Comment (vishesh92) [reply controllers/cks_cluster_controller.go:81]: We can't. That is why I have added a check in Line 77.

File: controllers/cks_machine_controller.go
Comment (vishesh92) [reply controllers/cks_machine_controller.go:68]: We can have a finalizer. But if the VM gets destroyed/expunged from ACS, it automatically gets removed from the CKS cluster as well.

File: api/v1beta3/cloudstackcluster_types.go
Comment (vishesh92) [reply api/v1beta3/cloudstackcluster_types.go:40]: As per my understanding we don't need to do that here. Pointer is used if we want to distinguish between explicit `false` and not specified.

Ref.: https://book.kubebuilder.io/cronjob-tutorial/api-design.html - search for `pointer` on the page.

File: controllers/cloudstackmachine_controller_test.go
Comment (vishesh92) [reply controllers/cloudstackmachine_controller_test.go:58]: We need it for the `CksClusterReconciler`.

**Comment by vishesh92** [comment]
@g-gaston I have addressed the comments.

File: api/v1beta3/cloudstackcluster_types.go
Comment (g-gaston) [reply api/v1beta3/cloudstackcluster_types.go:40]: I was talking about this: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

Again not required, but k8s does seem to recommend using pointer if the type doesnt have a built-in nil.

On the topic of needing to distinguish unset vs false:

> Using a pointer allows distinguishing unset from the zero value for that type. There are some cases where, in principle, a pointer is not needed for an optional field since the zero value is forbidden, and thus implies unset. There are examples of this in the codebase. However:
>
>*    it can be difficult for implementors to anticipate all cases where an empty value might need to be distinguished from a zero value
>*   structs are not omitted from encoder output even where omitempty is specified, which is messy;
>*   having a pointer consistently imply optional is clearer for users of the Go language client, and any other clients that use corresponding types

File: controllers/cks_cluster_controller.go
Comment (g-gaston) [inline controllers/cks_cluster_controller.go:80]: should this be an error? or maybe a no-op?
Why do we reconcile? it seems like the user would need to make changes for this to start working, right? Maybe I'm missing something

File: controllers/cks_machine_controller.go
Comment (g-gaston) [reply controllers/cks_machine_controller.go:68]: Ahh got you. Then do we even need the reconcile delete?

File: controllers/cloudstackmachine_controller_test.go
Comment (g-gaston) [reply controllers/cloudstackmachine_controller_test.go:58]: Could you help me understand this? I'm a bit lost. Why do we need to add a mock expectation on a test for one controller when the mock is being called in a different controller?

File: controllers/cks_machine_controller.go
Comment (vishesh92) [reply controllers/cks_machine_controller.go:68]: I added it in case the functionality upstream changes. So, this is good to have.

File: controllers/cloudstackmachine_controller_test.go
Comment (vishesh92) [reply controllers/cloudstackmachine_controller_test.go:58]: My bad. We don't need it. I mistook your comment for `cks_machine_controller_test.go`. removing this.

File: controllers/cks_cluster_controller.go
Comment (vishesh92) [reply controllers/cks_cluster_controller.go:80]: Makes sense. Let me do no-op here to match the behavior with cloudstack controller.

File: controllers/cks_machine_controller.go
Comment (g-gaston) [reply controllers/cks_machine_controller.go:68]: ok sounds good. but then i think we should have a finalizer
Not trying to be too purist, but if we are adding code to ensure the system works, even if not fully necessary right now, we should add it in a way that actually ensures the system works. If upstream changes and we don't have a finalizer, the cs vm can be deleted before we get a chance to execute this.

Comment (vishesh92) [reply controllers/cks_machine_controller.go:71]: @g-gaston I have added the finalizer.

File: api/v1beta3/cloudstackcluster_types.go
Comment (g-gaston) [reply api/v1beta3/cloudstackcluster_types.go:40]: @vishesh92 are you planning on changing this? or what's the decision here?
Otherwise the PR looks good to go

Comment (vishesh92) [reply api/v1beta3/cloudstackcluster_types.go:40]: I was thinking of keeping it as bool only.

Comment (vishesh92) [reply api/v1beta3/cloudstackcluster_types.go]: @g-gaston I have updated the PR to make it a pointer to a bool. Please check now.

**Review by g-gaston — COMMENTED** [review]
/lgtm

Thanks a lot!

**Comment by vishesh92** [comment]
> /lgtm
> 
> Thanks a lot!
> 
> I see you are trying to run tests. Let me know when you are done and ready to merge and I'll approve

Thanks @g-gaston for review. @kiranchavala is currently testing this PR. Once he gives a go ahead, we can merge this PR. Meanwhile, I will fix the conflicts in go.mod.

**Comment by rohityadavcloud** [comment]
@vishesh92 while I opened the PR, it’s largely your work. Pl feel free to resubmit a new PR or lgtm this when this is ready after QA. cc @kiranchavala

**Review by kiranchavala — APPROVED** [review]
LGTM,

Tested manually and 


A externally managed kubernetes  cluster got created in cloudstack when a capc deployment was initiated 

Screenshot 2024-06-24 at 4 07 04 PM

Screenshot 2024-06-24 at 4 07 33 PM

**Pull request merged**
